File: nodejs/plesk_api_client.js
// Copyright 1999-2016. Parallels IP Holdings GmbH. All Rights Reserved.
/* jshint -W069 */
"use strict";

const http = require('http');
const https = require('https');

class Client {

    constructor(host, port, protocol) {
        this._host = host;
        this._port = port || 8443;
        this._protocol = protocol || 'https';
    }

    setCredentials(login, password) {
        this._login = login;
        this._password = password;
    }

    setSecretKey(secretKey) {
        this._secretKey = secretKey;
    }

    request(body, callback) {
        let headers = {
            'Content-type': 'text/xml',
            'HTTP_PRETTY_PRINT': 'TRUE'
        };

        if (this._secretKey) {
            headers['KEY'] = this._secretKey;
        } else {
            headers['HTTP_AUTH_LOGIN'] = this._login;
            headers['HTTP_AUTH_PASSWD'] = this._password;
        }

        let options = {
            host: this._host,
            port: this._port,
            path: '/enterprise/control/agent.php',
            method: 'POST',
            headers: headers
        };

        let handler = (response) => {
            let result = '';
            response.on('data', (chunk) => result += chunk);
            response.on('end', () => callback(result));
        };

        let client = 'https' == this._protocol ? https : http;
        let request = client.request(options, handler);
        request.write(body);
        request.end();
    }

}

exports.Client = Client;

File: python/plesk_api_client.py
# Copyright 1999-2016. Parallels IP Holdings GmbH. All Rights Reserved.

import httplib
import ssl

class PleskApiClient:

    def __init__(self, host, port = 8443, protocol = 'https', ssl_unverified = False):
        self.host = host
        self.port = port
        self.protocol = protocol
        self.secret_key = None
        self.ssl_unverified = ssl_unverified

    def set_credentials(self, login, password):
        self.login = login
        self.password = password

    def set_secret_key(self, secret_key):
        self.secret_key = secret_key

    def request(self, request):
        headers = {}
        headers["Content-type"] = "text/xml"
        headers["HTTP_PRETTY_PRINT"] = "TRUE"

        if self.secret_key:
            headers["KEY"] = self.secret_key
        else:
            headers["HTTP_AUTH_LOGIN"] = self.login
            headers["HTTP_AUTH_PASSWD"] = self.password

        if 'https' == self.protocol:
            if self.ssl_unverified == True:
                conn = httplib.HTTPSConnection(self.host, self.port, context=ssl._create_unverified_context())
            else:
                conn = httplib.HTTPSConnection(self.host, self.port)
        else:
            conn = httplib.HTTPConnection(self.host, self.port)

        conn.request("POST", "/enterprise/control/agent.php", request, headers)
        response = conn.getresponse()
        data = response.read()
        return data
